### ISSUE TITLE
Google-fluentd stopped flushing logs to stackdriver (on April 8th)

### ISSUE THREAD

**Issue by bvandewalle** [GoogleCloudPlatform/fluent-plugin-google-cloud#393]
on a bunch of GCP instances, we discovered that google-fluentd stopped sending logs to Stackdriver on April 8th (possibly after this outage? https://status.cloud.google.com/incident/zall/20005)

Logging on the nodes and tailing the logs shows the following error:

```
$ sudo head /var/log/google-fluentd/google-fluentd.log
2020-04-10 06:25:01 +0000 [info]: flushing all buffer forcedly
2020-04-10 06:25:01 +0000 [info]: detected rotation of /var/log/syslog; waiting 5 seconds
2020-04-10 06:25:01 +0000 [info]: following tail of /var/log/syslog
2020-04-10 06:25:18 +0000 [warn]: failed to flush the buffer. retry_time=2 next_retry_seconds=2020-04-10 06:25:20 +0000 chunk="5a2e9a76141e92cd36b6d7a77beb3095" error_class=GRPC::Unavailable error="14:Deadline Exceeded"
  2020-04-10 06:25:18 +0000 [warn]: suppressed same stacktrace
2020-04-10 06:25:29 +0000 [warn]: failed to flush the buffer. retry_time=3 next_retry_seconds=2020-04-10 06:25:33 +0000 chunk="5a2e95f8ccb9ffb353a0265975bc50da" error_class=GRPC::Unavailable error="14:Deadline Exceeded"
  2020-04-10 06:25:29 +0000 [warn]: suppressed same stacktrace
2020-04-10 06:25:51 +0000 [error]: failed to flush the buffer, and hit limit for retries. dropping all chunks in the buffer queue. retry_times=3 records=628436 error_class=GRPC::Unavailable error="14:Deadline Exceeded"
  2020-04-10 06:25:51 +0000 [error]: suppressed same stacktrace
2020-04-10 06:25:54 +0000 [warn]: failed to flush the buffer. retry_time=0 next_retry_seconds=2020-04-10 06:25:55 +0000 chunk="5a2e985c83d03f95b2cf2414682016d2" error_class=GRPC::Unavailable error="14:Deadline Exceeded"
```

here is the stacktrace:
```
2020-04-08 14:27:41 +0000 [warn]: failed to flush the buffer. retry_time=1 next_retry_seconds=2020-04-08 14:27:42 +0000 chunk="5a2c82975293bdeac71fe2a4b6c383a2" error_class=GRPC::Unavailable error="14:Deadline Exceeded"
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:31:in `check_status'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:181:in `attach_status_results_and_complete_call'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:377:in `request_response'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:172:in `block (2 levels) in request_response'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:171:in `block in request_response'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:363:in `block in retryable'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:230:in `block in create_api_call'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:260:in `block in create_api_call'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-cloud-logging-1.6.6/lib/google/cloud/logging/v2/logging_service_v2_client.rb:525:in `write_log_entries'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.4/lib/fluent/plugin/out_google_cloud.rb:882:in `write_request_via_grpc'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.4/lib/fluent/plugin/out_google_cloud.rb:791:in `call'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.4/lib/fluent/plugin/out_google_cloud.rb:791:in `write'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/compat/output.rb:131:in `write'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:1125:in `try_flush'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:1431:in `flush_thread_run'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:461:in `block (2 levels) in start'
  2020-04-08 14:27:41 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
```

Restarting fluentd seems to make it work again. Any idea how can we prevent this from happening in the future ?

cc @bplotnick

**Comment by igorpeshansky** [comment]
This is indeed related to the outage you quoted. As far as we could see, the gRPC libraries get stuck when unable to process API authentication (because of some cache), and restarting the process is the only way to reset the cache. We are trying to reproduce in our test environment and investigate, but so far no luck.

**Comment by cablehead** [comment]
Adding a data point. We also encounter this traceback maybe once or twice every other month. 

Most recent occurrence is  `2020-05-27 17:00:26 +0000`

```
2020-05-27 17:00:26 +0000 [warn]: failed to flush the buffer. retry_time=0 next_retry_seconds=2020-05-27 17:00:27 +0000 chunk="5a6a401fae1b8050c0691dd2b8683cf6" error_class=GRPC::Unavailable error="14:Deadline Exceeded"
--
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:31:in `check_status'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:181:in `attach_status_results_and_complete_call'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:377:in `request_response'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:172:in `block (2 levels) in request_response'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/grpc-1.22.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:171:in `block in request_response'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:363:in `block in retryable'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:230:in `block in create_api_call'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-gax-1.7.1/lib/google/gax/api_callable.rb:260:in `block in create_api_call'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/google-cloud-logging-1.6.6/lib/google/cloud/logging/v2/logging_service_v2_client.rb:525:in `write_log_entries'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.5/lib/fluent/plugin/out_google_cloud.rb:882:in `write_request_via_grpc'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.5/lib/fluent/plugin/out_google_cloud.rb:791:in `call'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluent-plugin-google-cloud-0.8.5/lib/fluent/plugin/out_google_cloud.rb:791:in `write'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/compat/output.rb:131:in `write'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:1125:in `try_flush'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:1431:in `flush_thread_run'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin/output.rb:461:in `block (2 levels) in start'
  | 2020-05-27 17:00:26 +0000 [warn]: /opt/google-fluentd/embedded/lib/ruby/gems/2.6.0/gems/fluentd-1.7.4/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
```

Once google-fluentd encounters this traceback, the process stays up (so the pod won't restart), but no longer forwards any logs. A restart fixes the issue.

**Comment by igorpeshansky** [comment]
Update: this is fixed by grpc/grpc#19756 and grpc/grpc#23333. We have to wait until those fixes appear in the next gRPC release before we can pick them up for the `google-fluentd` package.

**Comment by igorpeshansky** [comment]
The fixes are backported to gRPC [1.30.1](https://github.com/grpc/grpc/releases/tag/v1.30.1).
